### DISCUSSION TITLE
Android Mediacodec encoder/decoder support

### DISCUSSION THREAD

**Discussion by igosoft** [baresip/baresip#2709] [Q&A]
Hi,

I have started working on enabling the Android Mediacodec **decoder** in the avcodec module (https://github.com/baresip/baresip/discussions/2245). After a few hours, I had function for gathering PPS and SPS data, which I use to start the decoder. However, it seems that some SPS and PPS data are out of range. That's why I decided to check if the Android Mediacodec encoder works. Unfortunately, it does not work properly - received video frames are small, and the video bitrate is very close to zero, around 20 kbit/s (Android camera works, frames are sent) 

Does anyone have experience with encoding video frames using h264_mediacodec and Android camera as source ?

Some important config file values:
video_size		640x480
video_bitrate		512000
video_fps		        30
video_fullscreen	no
videnc_format		yuv420p

video_source        avformat,android_camera,1
avcodec_h264enc h264_mediacodec
avcodec_h264dec h264_mediacodec
avcodec_hwaccel mediacodec

**Comment by frogblueDFi** [comment]
Hi Igor!

After a couple of month I resume work on trying to get the Android Mediacodec hardware acceleration to work.
I ended up earlier with results in topic #2245.

Audio and video is now ok running on most Android devices if video quality is not configured to high.
But for better performance I would like to use hw acceleration as I do with VideoToolbox for iOS/macOS.

So what I am missing is the gathering of the PPS and SPS data?
Your wrote that you have written functions for this.
Could you give me a hint how to do this or even provide code, so I can try MediaCodec encoder functionality in my app environment?

If we could bring it up and running, we could create a Pull Request.

Thanx a lot for any help or advice

Dietmar

**Comment by igosoft** [comment]
Hi Dietmar

It's WIP and I gave up at some point when I realized that I have too little experience with ffmepg and not enough spare time  
[hw_wip.zip](https://github.com/user-attachments/files/18816684/hw_wip.zip)

Please notice avcodec_open2 must be called only when SSP / PPS data have been decoded (in other words init_decoder function can't call avcodec_open2 )

**Reply by frogblueDFi** [reply]
Thanx a lot! Less experience and less time also on my side :-)
Will have a look at it, great help to start work again.
Just one question. You did only work on the decoding part. What about encoding?

**Comment by igosoft** [comment]
When it comes to encoding it's relatively easy to implement using Android
SDK. Encoded data can be then send using baresip/re.
Probably, decoding is also easier to implement in Android SDK then using ndk

pon., 17 lut 2025, 09:12 użytkownik frogblueDFi ***@***.***>
napisał:
